exporter/otlpexporter/otlp.go

```diff
@@ -0,0 +1,208 @@
+	errAlreadyStopped
+)
+
+// NewTraceExporter creates an OTLP trace exporter.
+func NewTraceExporter(logger *zap.Logger, config configmodels.Exporter) (exporter.TraceExporter, error) {
+	oce, err := createOTLPExporter(logger, config)
+	if err != nil {
+		return nil, err
+	}
+	oexp, err := exporterhelper.NewTraceExporter(
+		config,
+		oce.pushTraceData,
+		exporterhelper.WithShutdown(oce.Shutdown))
+	if err != nil {
+		return nil, err
+	}
+
+	return oexp, nil
+}
+
+// NewMetricsExporter creates an OTLP metrics exporter.
+func NewMetricsExporter(logger *zap.Logger, config configmodels.Exporter) (exporter.MetricsExporter, error) {
+	oce, err := createOTLPExporter(logger, config)
+	if err != nil {
+		return nil, err
+	}
+	_, err = exporterhelper.NewMetricsExporter(
+		config,
+		oce.pushMetricsData,
+		exporterhelper.WithShutdown(oce.Shutdown),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: return created exporter after metric support is finalized in PushMetricsData()
+	return nil, configerror.ErrDataTypeIsNotSupported
+}
+
+// createOTLPExporter creates an OTLP exporter.
+func createOTLPExporter(logger *zap.Logger, config configmodels.Exporter) (*otlpExporter, error) {
+	oCfg := config.(*Config)
+
+	if oCfg.Endpoint == "" {
+		return nil, &exporterError{
+			code: errEndpointRequired,
+			msg:  "OTLP exporter config requires an Endpoint",
+		}
+	}
+
+	numWorkers := defaultNumWorkers
+	if oCfg.NumWorkers > 0 {
+		numWorkers = oCfg.NumWorkers
+	}
+
+	exportersChan := make(chan *exporterImp, numWorkers)
+	for exporterIndex := 0; exporterIndex < numWorkers; exporterIndex++ {
+		// TODO: newExporter blocks for connection. Now that we have ability
+		// to report errors asynchronously using Host.ReportFatalError we can move this
+		// code to Start() and do it in background to avoid blocking Collector startup
+		// as we do now.
+		exporter, serr := newExporter(oCfg)
+		if serr != nil {
+			return nil, fmt.Errorf("cannot configure OTLP exporter: %v", serr)
+		}
+		exportersChan <- exporter
+	}
+	oce := &otlpExporter{exporters: exportersChan}
+	return oce, nil
+}
+
+func (oce *otlpExporter) Shutdown() error {
+	// Stop all exporters. Will wait until all are stopped.
+	wg := &sync.WaitGroup{}
+	var errors []error
+	var errorsMu sync.Mutex
+	visitedCnt := 0
+	for currExporter := range oce.exporters {
+		wg.Add(1)
+		go func(exporter *exporterImp) {
+			defer wg.Done()
+			err := exporter.stop()
+			if err != nil {
+				errorsMu.Lock()
+				errors = append(errors, err)
+				errorsMu.Unlock()
+			}
+		}(currExporter)
+		visitedCnt++
+		if visitedCnt == cap(oce.exporters) {
+			// Visited and concurrently executed stop() on all exporters.
+			break
+		}
+	}
+
+	// Wait for all stop() calls to finish.
+	wg.Wait()
+	close(oce.exporters)
+
+	return oterr.CombineErrors(errors)
+}
+
+func (oce *otlpExporter) pushTraceData(ctx context.Context, td consumerdata.TraceData) (int, error) {
+	// Get first available exporter.
+	exporter, ok := <-oce.exporters
```

Comment:
At some point we had talked about not doing workers in an exporter and instead simply exporting synchronously so that the pipeline would completely handle asynchronous sending on it's own. Not sure if we made the necessary changes that were needed (if any) to enable this. //cc @pjanotti

Reply:
We don't currently have a way to make this work. Concurrency is currently defined in OTLP and is an inherent part of OTLP exporter implementation. I doubt that it can be support in the pipeline in a way that is suitable for all concurrency patterns that may need to be needed by arbitrary protocols.